library/firewall_lib_facts.py

```diff
@@ -0,0 +1,136 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = """
+module: firewall_lib_facts
+short_description: module for firewall role
+requirements:
+  - python3-firewall or python-firewall
+description:
+  Get firewalld configuration information, including default settings
+  and custom settings. There are no options for this module
+author: Brennan Paciorek (@BrennanPaciorek)
+"""
+
+RETURN = """
+changed:
+  description:
+    whether anything was changed, most likely false
+  type: bool
+  returned: always
+  sample: False
+ansible_facts:
+  description:
+    any facts that are returned via this module running
+    currently only gets firewalld configuration
+    when calling the module
+  type: dict
+  returned: always
+  sample: {
+      "config": {
+          "default": {},
+          "custom": {},
+      },
+  }
+"""
+
+EXAMPLES = """
+# Run with no parameters to gather facts
+firewall_lib_facts:
+"""
+
+
+from ansible.module_utils.basic import AnsibleModule
+import os
+
+try:
+    import firewall.config
+
+    HAS_FIREWALLD = True
+except ImportError:
+    HAS_FIREWALLD = False
+
+
+def config_to_dict(module, fw=None):
+    config = {}
+    defaults = {}
+    custom = {}
+    setting_list = ["zones", "services", "icmptypes", "helpers", "ipsets"]
+
+    try:
+        firewall.config.FIREWALLD_POLICIES
+        setting_list.append("policies")
+    except AttributeError:
+        pass
+
+    for setting in setting_list:
+        default_setting_dir = os.path.join(firewall.config.USR_LIB_FIREWALLD, setting)
+        custom_setting_dir = os.path.join(firewall.config.ETC_FIREWALLD, setting)
+
+        defaults[setting] = fetch_settings_from_dir(default_setting_dir)
+        custom[setting] = fetch_settings_from_dir(custom_setting_dir)
+
+    config["default"] = defaults
+    config["custom"] = custom
+
+    return config
+
+
+def fetch_settings_from_dir(directory):
+    setting_options = []
+
+    for _file in os.listdir(directory):
+        is_setting = ".xml" in _file and ".old" not in _file
+        if is_setting:
+            setting_options.append(_file[:-4])
+
+    return setting_options
+
+
+def main():
+
+    # module_args = dict(
+    #    detailed=dict(type="bool", default=True, required=False)
+    # )
+
+    results = dict(changed=False, ansible_facts=dict())
+
+    module = AnsibleModule(argument_spec={}, supports_check_mode=True)
+
+    if not HAS_FIREWALLD:
+        module.fail_json(msg="firewalld not installed")
+
+    if module.check_mode:
+        module.exit_json(**results)
+
+    results["ansible_facts"]["firewalld_config"] = config_to_dict(module)
```

Comment:
Why is this being returned as `ansible_facts.firewalld_config` then you have to do
```yaml
  set_fact:
    firewall_config: "{{ __firewalld_facts.ansible_facts.firewalld_config }}"
```
to set `firewall_config`?  I guess you could shorten it by dropping the `ansible_facts` part

Reply:
Yeah, I added the ansible facts to account for any additional ansible facts that might be added later on, although I cannot think of any more that would be used for this role, so I can drop the ansible_facts part.